## 4.8 Grouping data & Aggregating variables¶

#### 01. Import libraries

In [5]:
# Import libraries
import pandas as pd
import numpy as np
import os

#### 02. Import dataset & Create a path

In [7]:
# path
path = r'/Users/renatabatista/Other Docs/Germany/CareerFoundry/Data Imersion/4. Python/14-04-2025 Instacart Basket Analysis'

In [8]:
# Import the ords_prods_merged_updated.pkl - DERIVED task
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_merged_new_variables_derived.pkl'))

In [9]:
# Check dimensions of imported dataframe for consistency
ords_prods_merge.shape

(32404859, 21)

#### 03. Grouping data with Pandas

##### In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.


In [11]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

#### 04. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

##### Compared to the results from the subset, the results from the entire dataframe includes ALL 'department_id' values

##### Additionally, the mean values for the department ids present in the subset have all changed.
##### Some have increased while others decreased

##### For example, in the subset department id 10 had a mean value of '18.681852', while in the entire dataframe it has a mean value of '20.197148'

#### 05. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

##### These operations were already performed on the dataframe during the exercise

In [15]:
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   
15    186761     1494             7                   2                  8   
16    317121     1494             9                   5                  8   
17    200390     1540             4                   2                  0   
18    374436     1540             5                   1                 21   
19   2920553     1540             6                   1                  0   

    ...  first_order  add_to_cart_order  reordered  _merge price_range_loc  \
0   ...        False                  5          0    both             NaN   
1   ...        False                  1          1    both             NaN   
2   ...        False                 20          0    both             NaN   
3   ...         True                 10          0    both             NaN   
4   ...        False                 11          1    both             NaN   
5   ...        False                  7          0    both             NaN   
6   ...        False                  2          0    both             NaN   
7   ...        False                  1          0    both             NaN   
8  

In [16]:
# Check loyalty flags
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

#### 06. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [18]:
# Use the 'groupby' function to split the data into groups based on the 'loyalty_flags' column
# Aggregate the data based on prices, providing basic statistics such as mean, min, max, std and median
# The results show the average prices of the products purchased by each category of customer (loyal, regular, new)
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'std', 'median']})

prices                                 
                       mean  min      max         std median
loyalty_flag                                                
Loyal customer    10.386336  1.0  99999.0  328.017787    7.4
New customer      13.294670  1.0  99999.0  597.560299    7.4
Regular customer  12.495717  1.0  99999.0  539.720919    7.4

#### 07.The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

##### If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

##### If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [20]:
# Create a new column for the average spending of a customer
ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [21]:
# Check the new column
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   
15    186761     1494             7                   2                  8   
16    317121     1494             9                   5                  8   
17    200390     1540             4                   2                  0   
18    374436     1540             5                   1                 21   
19   2920553     1540             6                   1                  0   

    ...  add_to_cart_order  reordered  _merge  price_range_loc  \
0   ...                  5          0    both              NaN   
1   ...                  1          1    both              NaN   
2   ...                 20          0    both              NaN   
3   ...                 10          0    both              NaN   
4   ...                 11          1    both              NaN   
5   ...                  7          0    both              NaN   
6   ...                  2          0    both              NaN   
7   ...                  1          0    both              NaN   
8   ...                  1          0    both              NaN   
9   ...                  2          1    both

#### Assign spender tag to each customer

In [23]:
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

In [24]:
ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [25]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

##### The results show a relatively low number of high spenders, once again, likely due to the removal of outliers

#### 08. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

##### If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”

##### If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”

##### If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [28]:
# Create a new column for customer frequency flags based on the median values from 'days_since_prior_order' column
ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [29]:
# Check that the column was created successfully and properly
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   
15    186761     1494             7                   2                  8   
16    317121     1494             9                   5                  8   
17    200390     1540             4                   2                  0   
18    374436     1540             5                   1                 21   
19   2920553     1540             6                   1                  0   

    ...  _merge  price_range_loc     busiest_day    busiest_days  \
0   ...    both              NaN  Regularly busy  Regularly busy   
1   ...    both              NaN  Regularly busy  Regularly busy   
2   ...    both              NaN     Busiest day    Busiest days   
3   ...    both              NaN  Regularly busy    Slowest days   
4   ...    both              NaN      Least busy    Slowest days   
5   ...    both              NaN  Regularly busy    Busiest days   
6   ...    both              NaN  Regularly busy  Regularly busy   
7   ...    both              NaN  Regularly busy    Slowest days   
8   ...    both              NaN  Regularly busy    Busiest days   
9   ...    both          

#### Based on the criteria provided, create a column featuring order frequency flags

In [31]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [32]:
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] > 10) & (ords_prods_merge['customer_frequency'] <=20), 'frequency_flag'] = 'Regular customer'

In [33]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [34]:
# Check value counts for new column
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [35]:
ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(60)

user_id  customer_frequency         frequency_flag
0       138                 8.0      Frequent customer
1       138                 8.0      Frequent customer
2       709                 8.0      Frequent customer
3       764                 9.0      Frequent customer
4       764                 9.0      Frequent customer
5       777                11.0       Regular customer
6       825                20.0       Regular customer
7       910                 6.0      Frequent customer
8      1052                10.0      Frequent customer
9      1052                10.0      Frequent customer
10     1379                20.0       Regular customer
11     1480                 7.0      Frequent customer
12     1480                 7.0      Frequent customer
13     1480                 7.0      Frequent customer
14     1494                 7.0      Frequent customer
15     1494                 7.0      Frequent customer
16     1494                 7.0      Frequent customer
17     1540                 7.0      Frequent customer
18     1540                 7.0      Frequent customer
19     1540                 7.0      Frequent customer
20     1540                 7.0      Frequent customer
21     1540                 7.0      Frequent customer
22     1540                 7.0      Frequent customer
23     1540                 7.0      Frequent customer
24     1540                 7.0      Frequent customer
25     1540                 7.0      Frequent customer
26     1540                 7.0      Frequent customer
27     1540                 7.0      Frequent customer
28     1540                 7.0      Frequent customer
29     1540                 7.0      Frequent customer
30     1540                 7.0      Frequent customer
31     1540                 7.0      Frequent customer
32     1540                 7.0      Frequent customer
33     1540                 7.0      Frequent customer
34     1598                 2.0      Frequent customer
35     1598                 2.0      Frequent customer
36     1598                 2.0      Frequent customer
37     1647                25.0  Non-frequent customer
38     1647                25.0  Non-frequent customer
39     1647                25.0  Non-frequent customer
40     1647                25.0  Non-frequent customer
41     1647                25.0  Non-frequent customer
42     2833                13.0       Regular customer
43     2850                 7.0      Frequent customer
44     2857                11.0       Regular customer
45     3010                 7.0      Frequent customer
46     3029                28.0  Non-frequent customer
47     3395                18.0       Regular customer
48     3904                21.0  Non-frequent customer
49     4006                 7.0      Frequent customer
50     4006                 7.0      Frequent customer
51     4006                 7.0      Frequent customer
52     4006                 7.0      Frequent customer
53     4006                 7.0      Frequent customer
54     4122                 7.0      Frequent customer
55     4141                30.0  Non-frequent customer
56     4343                30.0  Non-frequent customer
57     4553                30.0  Non-frequent customer
58     4845                20.0       Regular customer
59     6866                30.0  Non-frequent customer

#### 09. Export your dataframe as a pickle file and store it correctly in your 'Prepared Data' folder

In [37]:
# Check shape before exporting
ords_prods_merge.shape

(32404859, 25)

In [ ]:
# Export the updated dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined_merged_new_variables_derived_aggregated.pkl'))

In [ ]:
ords_prods_merge.column